service/worker/batcher/activities.go

```diff
@@ -308,3 +332,85 @@ func isDone(ctx context.Context) bool {
 		return false
 	}
 }
+
+func getResetEventIDByType(ctx context.Context, resetType enumspb.ResetType, namespace string, workflowExecution *commonpb.WorkflowExecution, frontendClient workflowservice.WorkflowServiceClient) (int64, error) {
+	switch resetType {
+	case enumspb.RESET_TYPE_FIRST_WORKFLOW_TASK:
+		return getFirstWorkflowTaskEventID(ctx, resetType, namespace, workflowExecution, frontendClient)
+	case enumspb.RESET_TYPE_LAST_WORKFLOW_TASK:
+		return getLastWorkflowTaskEventID(ctx, resetType, namespace, workflowExecution, frontendClient)
+	default:
+		errorMsg := fmt.Sprintf("provided reset type (%v) is not supported.", resetType)
+		return 0, serviceerror.NewInvalidArgument(errorMsg)
+	}
+}
+
+func getLastWorkflowTaskEventID(ctx context.Context, resetType enumspb.ResetType, namespace string, workflowExecution *commonpb.WorkflowExecution, frontendClient workflowservice.WorkflowServiceClient) (workflowTaskEventID int64, err error) {
+	req := &workflowservice.GetWorkflowExecutionHistoryRequest{
+		Namespace:       namespace,
+		Execution:       workflowExecution,
+		MaximumPageSize: 1000,
+		NextPageToken:   nil,
+	}
+	for {
+		resp, err := frontendClient.GetWorkflowExecutionHistory(ctx, req)
+		if err != nil {
+			//TODO: What type of error should we return here?
+			return 0, errors.New("GetWorkflowExecutionHistory failed")
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
```

Comment:
I think we should reset to the corresponding scheduled event (which can be find from the event attribute of this completed one).

Reply:
Agree on the idea. But double check the event ID pass into the reset API is inclusive or exclusive.

Reply:
based on the comments in the code, this is exclusive.

Reply:
I actually don't see the benefits of resetting to the scheduled event. In both cases, the new workflow history will have the workflow task failed event. Resetting to completed can actually reuse more batches of events.